plugins/woocommerce/tests/php/src/Internal/Admin/BlockTemplates/BlockTemplateTest.php

```diff
@@ -35,6 +36,137 @@ public function test_add_block() {
 		$this->assertSame( $block, $template->get_block( 'test-block-id' ) );
 	}
 
+	/**
+	 * Test the after_add_block hooks fires.
+	 */
+	public function test_after_add_block_hooks() {
+		$template = new BlockTemplate();
+
+		$hook_called = false;
+
+		$after_add_block_hook = function( BlockInterface $block ) use ( &$hook_called ) {
+			$hook_called = true;
+
+			if ( 'test-block-id' === $block->get_id() ) {
+				$hook_called = true;
+			}
+
+			$root_template = $block->get_root_template();
+
+			if ( $root_template->get_block( 'test-block-id-2' ) ) {
+				// The block was already added, so just return.
+				// This short-circuiting done because this hook will be called two times:
+				// 1. When the `test-block-id` block is added to the root template.
+				// 2. When the `test-block-id-2` block is added to the template in this hook.
+				// Without this short-circuiting, the second time `add_block` is called in this
+				// hook would throw an exception, which is handled by the API (an error gets logged).
+				return;
+			}
+
+			$root_template->add_block(
+				[
+					'id'        => 'test-block-id-2',
+					'blockName' => 'test-block-name-2',
+				]
+			);
+		};
+
+		add_action( 'woocommerce_block_template_after_add_block', $after_add_block_hook );
+
+		$specific_hook_called = false;
+
+		$specific_after_add_block_hook = function( BlockInterface $block ) use ( &$specific_hook_called ) {
+			if ( 'test-block-id' === $block->get_id() ) {
+				$specific_hook_called = true;
+			}
+		};
+
+		add_action( 'woocommerce_block_template_area_uncategorized_after_add_block_test-block-id', $specific_after_add_block_hook );
+
+		$template->add_block(
+			[
+				'id'        => 'test-block-id',
+				'blockName' => 'test-block-name',
+			]
+		);
+
+		$this->assertTrue(
+			$hook_called,
+			'Failed asserting that that the hook was called.'
+		);
+
+		$this->assertTrue(
+			$specific_hook_called,
+			'Failed asserting that that the specific hook was called.'
+		);
+
+		$this->assertNotNull(
+			$template->get_block( 'test-block-id-2' ),
+			'Failed asserting that the block was added to the template from the hook.'
+		);
+
+		remove_action( 'woocommerce_block_template_after_add_block', $after_add_block_hook );
```

Comment:
Isn't there a chance this could not run if something goes sideways during the test, affecting other tests? If so, is there any chance we can help prevent that?

I see that this particular action is not used in any other tests. I'm just wondering. Feel free to leave it like this if you don't think it's a problem.

Reply:
Wrapped in try/finally